File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/types/message/ReadMemberAttributeResponse.java
package org.opcfoundation.webserver.types.message;

import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;
import org.eclipse.milo.opcua.stack.core.types.builtin.Variant;
import org.eclipse.milo.opcua.stack.core.types.enumerated.NodeClass;
import org.jspecify.annotations.Nullable;
import org.opcfoundation.webserver.addressspace.nodes.UaInstanceNode;
import org.opcfoundation.webserver.addressspace.nodes.UaVariable;

public class ReadMemberAttributeResponse {
    private final NodeClass nodeClass;
    private final String browseName;
    private final LocalizedText displayName;
    private final LocalizedText description;
    private final @Nullable NodeId dataTypeId;
    private final @Nullable Integer valueRank;
    private final @Nullable Integer accessLevel;
    private final @Nullable Boolean historizing;
    private final @Nullable Variant value;

    public ReadMemberAttributeResponse(
            NodeClass nodeClass,
            String browseName,
            LocalizedText displayName,
            LocalizedText description,
            @Nullable NodeId dataTypeId,
            @Nullable Integer valueRank,
            @Nullable Integer accessLevel,
            @Nullable Boolean historizing,
            @Nullable Variant value)
    {
        this.nodeClass = nodeClass;
        this.browseName = browseName;
        this.displayName = displayName;
        this.description = description;
        this.dataTypeId = dataTypeId;
        this.valueRank = valueRank;
        this.accessLevel = accessLevel;
        this.historizing = historizing;
        this.value = value;
    }

    public ReadMemberAttributeResponse(UaInstanceNode instanceDeclaration)
    {
        this.nodeClass = instanceDeclaration.nodeClass();
        this.browseName = instanceDeclaration.browseName();
        this.displayName = instanceDeclaration.displayName();
        this.description = instanceDeclaration.description();

        if (NodeClass.Variable == instanceDeclaration.nodeClass()) {
            UaVariable variable = (UaVariable)instanceDeclaration;
            this.dataTypeId = variable.dataType();
            this.valueRank = variable.valueRank();
            this.accessLevel = variable.accessLevel();
            this.historizing = variable.historizing();
            this.value = variable.value();
        } else {
            this.dataTypeId = null;
            this.valueRank = null;
            this.accessLevel = null;
            this.historizing = null;
            this.value = null;
        }
    }

    public NodeClass getNodeClass() {
        return nodeClass;
    }

    public String getBrowseName() {
        return browseName;
    }

    public LocalizedText getDisplayName() {
        return displayName;
    }

    public LocalizedText getDescription() {
        return description;
    }

    public @Nullable NodeId getDataTypeId() {
        return dataTypeId;
    }

    public @Nullable Integer getValueRank() {
        return valueRank;
    }

    public @Nullable Integer getAccessLevel() {
        return accessLevel;
    }

    public @Nullable Boolean getHistorizing() {
        return historizing;
    }

    public @Nullable Variant getValue() {
        return value;
    }
}

File: opcua-webservice-node/src/server/service/message/reactiveobject/ReadMemberAttributeResponse.ts
import { NodeClass } from "opcua-webapi";
import { UaLocalizedText, UaNodeId, UaVariant } from "opcua-webapi-ts";
import { UaInstanceNode } from "../../../addressspace/nodes/UaInstanceNode";
import { UaVariable } from "../../../addressspace/nodes/UaVariable";

export class ReadMemberAttributeResponse {
    private readonly _nodeClass: NodeClass;
    private readonly _browseName: string;
    private readonly _displayName: UaLocalizedText;
    private readonly _description: UaLocalizedText;
    private readonly _dataTypeId: UaNodeId | null;
    private readonly _valueRank: number | null;
    private readonly _accessLevel: number | null;
    private readonly _historizing: boolean | null;
    private readonly _value: UaVariant | null;

    constructor(
        nodeClass: NodeClass,
        browseName: string,
        displayName: UaLocalizedText,
        description: UaLocalizedText,
        dataTypeId: UaNodeId | null,
        valueRank: number | null,
        accessLevel: number | null,
        historizing: boolean | null,
        value: UaVariant | null,
    ) {
        this._nodeClass = nodeClass;
        this._browseName = browseName;
        this._displayName = displayName;
        this._description = description;
        this._dataTypeId = dataTypeId;
        this._valueRank = valueRank;
        this._accessLevel = accessLevel;
        this._historizing = historizing;
        this._value = value;
    }

    get nodeClass(): NodeClass {
        return this._nodeClass;
    }

    get browseName(): string {
        return this._browseName;
    }

    get displayName(): UaLocalizedText {
        return this._displayName;
    }

    get description(): UaLocalizedText {
        return this._description;
    }

    get dataTypeId(): UaNodeId | null {
        return this._dataTypeId;
    }

    get valueRank(): number | null {
        return this._valueRank;
    }

    get accessLevel(): number | null {
        return this._accessLevel;
    }

    get historizing(): boolean | null {
        return this._historizing;
    }

    get value(): UaVariant | null {
        return this._value;
    }

    static fromInstanceDeclaration(instanceDeclaration: UaInstanceNode): ReadMemberAttributeResponse {
        if (instanceDeclaration.nodeClass === NodeClass.Variable) {
            const variable = instanceDeclaration as UaVariable;

            return new ReadMemberAttributeResponse(
                instanceDeclaration.nodeClass,
                instanceDeclaration.browseName,
                instanceDeclaration.displayName,
                instanceDeclaration.description,
                variable.dataType,
                variable.valueRank,
                variable.accessLevel,
                variable.historizing,
                variable.value,
            );
        }

        return new ReadMemberAttributeResponse(
            instanceDeclaration.nodeClass,
            instanceDeclaration.browseName,
            instanceDeclaration.displayName,
            instanceDeclaration.description,
            null,
            null,
            null,
            null,
            null,
        );
    }
}